File: universitybackend/src/main/java/com/miraz/universitybackend/services/StudentService.java
package com.miraz.universitybackend.services;

import com.miraz.universitybackend.models.Student;
import com.miraz.universitybackend.repository.StudentRepo;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;


@Component

public class StudentService {

    @Autowired
    private StudentRepo repo;

    //add or update student info
    public Student addOrUpdateStudent(Student student){
        repo.save(student);
        return student;
    }

    // return all the student of the database
    public List<Student> getAllStudents(){
        return  repo.findAll();
    }

    // delete a student from the database
    public Optional<Student> deleteStudent(int id){
        Optional<Student> student = repo.findById(id);
        repo.deleteById(id);
        return student;
    }

    //find a student with their assign id
    public Optional<Student> findById(int id){
        return repo.findById(id);
    }

    // find a student with their name
    public List<Optional<Student>> findByName(String name){
        return repo.findByFirstNameStartingWithIgnoreCase(name);
    }
}

File: universityGui/src/app/services/student.service.ts
import { Injectable } from '@angular/core';
import{HttpClient} from '@angular/common/http'
import { Student} from '../models/student';
@Injectable({
  providedIn: 'root'
})
export class StudentService {

  api = "http://localhost:8080/student"
  constructor(private http: HttpClient) { }

  addOrUpdateStudent(student:Student){
    return this.http.post(this.api+"/add", student);
  }

  getAllStudents(){
    return this.http.get(this.api+"/find/all");
  }

  deleteStudent(id:number){
    return this.http.delete(this.api+"/delete/"+id);
  }

  updateStudent(student:Student){
    return this.http.put(this.api+"/update", student);
  }

  searchById(id:number){
    return this.http.get(this.api+"/find/id/"+id);

  }

  searchByName(name:String){
    return this.http.get(this.api+"/find/name/"+name);
  }
}
